### ISSUE TITLE
Glitches caused by OpenSL ES with wrong buffer size?

### ISSUE THREAD

**Issue by mrjohs** [libpd/pd-for-android#72]
Glitches caused by OpenSL ES with wrong buffer sizes?

I am building an app with a relatively large pd patch. It was working great on my Nexus 5 until I converted the patch to stereo, which led to twice of lots of things. Then glitches started happening. I do not think it is a performance issue, as the app runs fine on older hardware like the Samsung Galaxy Tab 10.1 (2011) and the Galaxy SIII.

I came across this article: [Understanding Android audio towards achieving low latency response](https://github.com/igorski/MWEngine/wiki/Understanding-Android-audio-towards-achieving-low-latency-response#user-content-the-right-buffer-size)

*duh. As mentioned before: larger values improve stability, while lower values lower latency. However, it's not just a matter of choosing the right size for the minimum amount of latency and maximum stability, but also making it a multiple of the devices native buffer size.*

*For instance : some devices may report a recommended 4800 samples per buffer-size, which is apart from being quite large, also an unusual number. By dividing using multiples of 4800 we can however reach a low, stable buffer size of 75 samples per buffer. Other devices may however report a recommended buffer size of 512 samples, the lowest usable buffer size ( when keeping the above rule of multiples in mind ) could be 64 samples. Using a value outside of the multiple range ( for instance using 75 samples per buffer on a device that requires 64 samples per buffer ) may cause glitches as occasionally the buffer callback is called twice per timeslice. This can go by unnoticed if you have CPU cycles to spare, but it's more likely that this will slowly but surely accumulate to a clusterfuck.*

This rhymes very well with my experience.

When running libpd with opensl it seems like the buffer size always is 64. 

When I do suggestOutputBufferSize(suggestSampleRate()) on the Nexus 5 I get 240. 240 is not a multiple of 64, and I get glitches. On a Honor 7 that I have tried, suggestOutputBufferSize(suggestSampleRate()) returns 960, which IS a multiple of 64, and there are no glitches. On the Galaxy SIII suggestOutputBufferSize(suggestSampleRate()) returns 1024, which is also a multiple of 64, and no glitches there either.

If I run the app at 44100 on the Nexus 5 there are no glitches (but AUDIO_OUTPUT_FLAG_FAST denied by client), but then there are glitches on the Honor 7. The Galaxy SIII is fine with both 44100 and 48000.

So my question is:

Could it be that the buffer size for opensl should be a multiple of what is returned by suggestOutputBufferSize(suggestSampleRate()) to achieve optimal performance? If yes: Would it be possible to modify pd-for-android, so that the audio buffer size could be set manually?

I realise that libpd can only process n ticks at a time, and that those should rather be 64 samples each. So in case of buffer size 240 one would have to process 4 ticks, which would end up with 256 samples, then send 240 of them to the soundcard and save 16 of them for the next cycle. It would be a bit complicated, but I have dug a bit around in z_jni_opensl.c and it seems like it should be possible.

But before I make the effort: Do you think it will solve the problem of glitches?

Thanks for any thoughts.

**Comment by mrjohs** [comment]
I don't think the issue occurs because of the buffer size. I hacked together a sketch of my proposal for buffer size of 240, and the problem persisted. I also tried the workaround with fake touches. This solved the problem in a large degree, but not completely, and it introduced other (well known) issues.

Also, I found that the Nexus 7, which also has a buffer size of 240, does not produce glitches.

Rather, I think this is the same issue as #47.
